### PR TITLE
20 & 21 - skip metadata pipeline

### LINKED ISSUE #20: Make data-providers *Handlers use Metadata present in Item without transformation
Dataverse provides the content of the `<record><metadata>` tag directly from a cached export. We need to be able to expose these directly, without jumping through the loops of the XSL pipelines transformation.

As we expect to have VERY large metadata coming from DDI codebooks etc (easily >10MB), it is mandatory to keep this use case in mind for the refactoring.

Current idea is to change the `Metadata(InputStream)` to store the stream and copy it chunk by chunk into the XML writer when the element gets written.

### LINKED ISSUE #21: Make data-provider repositories aware of the metadata type requested
This is related to #20.

To make the underlying repository stuff the right metadata into the `Item`, we need to extend the repositories `getItem` with a way to know the circumstances.

Ideas:
1. Send the prefix as string?
2. Send `Context`?
3. Send `MetadataFormat` :heavy_check_mark:

### PR DESCRIPTION
**Pull request by poikilotherm**
Closes #20 
Closes #21

TODOs
- [x] Add tests for Metadata methods
- [x] Add tests for InMemoryItemRepo & GetRecordHandler & ListRecordsHandler with CopyElement backing

### CODE REVIEW

**Comment by poikilotherm** [comment]
@landreev  can you please review?

**Comment by poikilotherm** [comment]
@landreev  can you please review?

File: xoai-common/src/main/java/io/gdcc/xoai/util/ReplacingInputStream.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+/**
+ * Simple FilterInputStream that can replace occurrances of bytes with something else.
+ *
+ * Source copied from https://github.com/mwiesiolek/inbot-utils/blob/master/src/main/java/io/inbot/utils/ReplacingInputStream.java
+ * Originally licensed under MIT license
+ * Slightly adapted to avoid Apache Commons Lang3 dependency
+ *
+ * @author jillesvangurp
+ * @deprecated This should not be used anywhere - it's very slow and of limited use for XOAI
```
Comment (landreev) [inline xoai-common/src/main/java/io/gdcc/xoai/util/ReplacingInputStream.java]: Would it make sense to add another sentence here, spelling out that it's here for benchmarking/comparison only? - I was momentarily confused seeing it in the PR.

File: xoai-common/src/main/java/io/gdcc/xoai/xml/CopyElement.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            throw new XmlWriteException(e);
+        }
+    }
+    
+    /**
+     * A matcher, created only once, reusable to match the XML declaration with any attributes. Non-greedy, so
+     * we do not interfere with any XML processing instructions following.
+     */
+    private static final Matcher xmlDeclaration = Pattern.compile("<\\?xml .*?\\?>").matcher("");
+
```
Comment (landreev) [inline xoai-common/src/main/java/io/gdcc/xoai/xml/CopyElement.java:68]: is `<\\?xml[^>]*\\?>`  any better/prettier?

**Review by landreev — APPROVED** [review]
Super impressive work! 
Merge it merge it merge it :smile:

File: xoai-common/src/main/java/io/gdcc/xoai/util/ReplacingInputStream.java (hunk 1 above)
Comment (poikilotherm) [reply xoai-common/src/main/java/io/gdcc/xoai/util/ReplacingInputStream.java]: Yes, absolutely.

I will also move it from `main` to `test`, so no one uses it accidentally.

File: xoai-common/src/main/java/io/gdcc/xoai/xml/CopyElement.java (hunk 2 above)
Comment (poikilotherm) [reply xoai-common/src/main/java/io/gdcc/xoai/xml/CopyElement.java:68]: Let's use a saying here: "Geschmackssache", sagte der Affe und bis in die Schmierseife. ("Matter of taste", monkey said and took a bite from the soap.)

Not sure which is faster, both work.

File: xoai-common/src/main/java/io/gdcc/xoai/util/ReplacingInputStream.java (hunk 1 above)
Comment (poikilotherm) [reply xoai-common/src/main/java/io/gdcc/xoai/util/ReplacingInputStream.java]: Done with f131d30

**Pull request merged**